// athenaenv-extension-plugin/index.js
// src/ts-plugin/index.js
/** @type {import('typescript/lib/tsserverlibrary')} */
let ts;

function createFFloatRanges(text) {
  // decimal with optional underscores, optional fraction, optional exponent, then trailing 'f'
  const re = /(?<![\w$])(?:\d[\d_]*(?:\.\d[\d_]*)?|\.\d[\d_]*)(?:[eE][+\-]?\d[\d_]*)?f(?![\w$])/g;
  const ranges = [];
  let m;
  while ((m = re.exec(text)) !== null) {
    ranges.push({ start: m.index, end: m.index + m[0].length });
  }
  return ranges;
}

function overlaps(aStart, aLen, bStart, bEnd) {
  const aEnd = aStart + (aLen ?? 0);
  return !(aEnd <= bStart || aStart >= bEnd);
}

function fileIsJSLike(fileName) {
  return fileName.endsWith(".js") || fileName.endsWith(".jsx");
}

function pluginCreate(info) {
  // Log once so you can confirm the plugin is active (TypeScript: Open TS Server Log)
  try {
    info.project.projectService.logger.info("[athenaenv-extension-plugin] loaded");
  } catch {}

  const ls = info.languageService;

  function filterDiagnosticsForFile(fileName, diags) {
    try {
      if (!fileIsJSLike(fileName)) return diags;

      const program = ls.getProgram?.();
      const sf = program?.getSourceFile(fileName);
      if (!sf) return diags;

      const text = sf.getFullText();
      const ranges = createFFloatRanges(text);
      if (!ranges.length) return diags;

      return diags.filter(d => {
        const start = d.start ?? 0;
        const length = d.length ?? 0;
        for (const r of ranges) {
          if (overlaps(start, length, r.start, r.end)) {
            return false; // drop diagnostics that overlap trailing-f numbers
          }
        }
        return true;
      });
    } catch {
      return diags; // fail-safe: never suppress unrelated diagnostics
    }
  }

  const proxy = Object.create(null);
  for (const k of Object.keys(ls)) {
    proxy[k] = (...args) => ls[k](...args);
  }

  proxy.getSyntacticDiagnostics = (fileName) =>
    filterDiagnosticsForFile(fileName, ls.getSyntacticDiagnostics(fileName));

  proxy.getSemanticDiagnostics = (fileName) =>
    filterDiagnosticsForFile(fileName, ls.getSemanticDiagnostics(fileName));

  proxy.getSuggestionDiagnostics = (fileName) => {
    const diags = ls.getSuggestionDiagnostics?.(fileName) || [];
    return filterDiagnosticsForFile(fileName, diags);
  };

  return proxy;
}

/** @type {import('typescript/lib/tsserverlibrary').PluginModuleFactory} */
function init(mod) {
  ts = mod;
  return { create: pluginCreate };
}

module.exports = init;
